File: software/mosek/7/tools/examples/c/case_portfolio_1.c
/*
  File : case_portfolio_1.c

  Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.

  Description :  Implements a basic portfolio optimization model.
 */

#include <math.h>
#include <stdio.h>

#include "mosek.h"

#define MOSEKCALL(_r,_call)  ( (_r)==MSK_RES_OK ? ( (_r) = (_call) ) : ( (_r) = (_r) ) );

static void MSKAPI printstr(void *handle,
                            MSKCONST char str[])
{
  printf("%s",str);
} /* printstr */

int main(int argc, const char  argv[])
{
  char            buf[128];
  const MSKint32t n=3;
  const double    gamma=0.05,
                  mu[]={0.1073,  0.0737,  0.0627},
                  GT[][3]={{0.1667,  0.0232,  0.0013},
                           {0.0000,  0.1033, -0.0022},
                           {0.0000,  0.0000,  0.0338}},
                  x0[3]={0.0, 0.0, 0.0},
                  w=1.0;
  double          rtemp;
  MSKenv_t        env;                        
  MSKint32t       k,i,j,offsetx,offsets,offsett,*sub;                     
  MSKrescodee     res=MSK_RES_OK;                   
  MSKtask_t       task; 

  sub = (MSKint32t *) calloc(n,sizeof(MSKint32t));

  res = sub==NULL ? MSK_RES_ERR_SPACE : MSK_RES_OK;

  /* Initial setup. */ 
  MOSEKCALL(res,MSK_makeenv(&env,NULL));
  MOSEKCALL(res,MSK_maketask(env,0,0,&task));
  MOSEKCALL(res,MSK_linkfunctotaskstream(task,MSK_STREAM_LOG,NULL,printstr));

  rtemp = w;
  for(j=0; j<n; ++j)
    rtemp += x0[j];

  /* Constraints. */ 
  MOSEKCALL(res,MSK_appendcons(task,1+n));                       
  MOSEKCALL(res,MSK_putconbound(task,0,MSK_BK_FX,rtemp,rtemp));
  sprintf(buf,"%s","budget");
  MOSEKCALL(res,MSK_putconname(task,0,buf));

  for(i=0; i<n; ++i)
  {
    MOSEKCALL(res,MSK_putconbound(task,1+i,MSK_BK_FX,0.0,0.0));
    sprintf(buf,"GT[%d]",1+i);
    MOSEKCALL(res,MSK_putconname(task,1+i,buf));
  }

  /* Variables. */
  MOSEKCALL(res,MSK_appendvars(task,1+2*n));  

  offsetx = 0;   /* Offset of variable x into the API variable. */
  offsets = n;   /* Offset of variable x into the API variable. */
  offsett = n+1; /* Offset of variable t into the API variable. */

  /* x variables. */
  for(j=0; j<n; ++j)  
  {
    MOSEKCALL(res,MSK_putcj(task,offsetx+j,mu[j]));
    MOSEKCALL(res,MSK_putaij(task,0,offsetx+j,1.0));
    for(k=0; k<n; ++k)	
      if( GT[k][j]!=0.0 )
        MOSEKCALL(res, MSK_putaij(task, 1 + k, offsetx + j, GT[k][j])); 

    MOSEKCALL(res,MSK_putvarbound(task,offsetx+j,MSK_BK_LO,0.0,MSK_INFINITY));	 
    sprintf(buf,"x[%d]",1+j);
    MOSEKCALL(res,MSK_putvarname(task,offsetx+j,buf));
  }

  /* s variable. */
  MOSEKCALL(res,MSK_putvarbound(task,offsets+0,MSK_BK_FX,gamma,gamma));
  sprintf(buf,"s");
  MOSEKCALL(res,MSK_putvarname(task,offsets+0,buf));

  /* t variables. */
  for(j=0; j<n; ++j)  
  {
    MOSEKCALL(res,MSK_putaij(task,1+j,offsett+j,-1.0));
    MOSEKCALL(res,MSK_putvarbound(task,offsett+j,MSK_BK_FR,-MSK_INFINITY,MSK_INFINITY));	 
    sprintf(buf,"t[%d]",1+j);
    MOSEKCALL(res,MSK_putvarname(task,offsett+j,buf));
  }
 
  sub[0] = offsets+0; 
  for(j=0; j<n; ++j)
    sub[j+1] = offsett+j;

  MOSEKCALL(res,MSK_appendcone(task,MSK_CT_QUAD,0.0,n+1,sub));
  MOSEKCALL(res,MSK_putconename(task,0,"stddev"));

  MOSEKCALL(res,MSK_putobjsense(task,MSK_OBJECTIVE_SENSE_MAXIMIZE));

  #if 0
  /* Turn all logout put off. */
  MOSEKCALL(res,MSK_putintparam(task,MSK_IPAR_LOG,0));
  #endif
   
  #if 0
  /* Dump the problem to a human readable OPF file. */ 
  MOSEKCALL(res,MSK_writedata(task,"dump.opf")); 
  #endif
    
  MOSEKCALL(res,MSK_optimize(task));

  #if 1
  /* Display the solution summary for quick inspection of results. */
  MSK_solutionsummary(task,MSK_STREAM_MSG);
  #endif

  if ( res==MSK_RES_OK )
  {
    double expret=0.0,stddev=0.0,xj;

    for(j=0; j<n; ++j)  	
    {
      MOSEKCALL(res,MSK_getxxslice(task,MSK_SOL_ITR,offsetx+j,offsetx+j+1,&xj));
      expret += mu[j]*xj;
    }  	

    MOSEKCALL(res,MSK_getxxslice(task,MSK_SOL_ITR,offsets+0,offsets+1,&stddev));

    printf("\nExpected return %e for gamma %e\n",expret,stddev); 	
  }	

  free(sub);

  return ( 0 );
}  

File: software/mosek/7/tools/examples/dotnet/case_portfolio_1.cs
using System;

/*
  File : case_portfolio_1.cs

  Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.

  Description :  Implements a basic portfolio optimization model.
*/


class msgclass : mosek.Stream  
{ 
  string prefix; 
  public msgclass (string prfx)  
  { 
    prefix = prfx; 
  } 
   
  public override void streamCB (string msg) 
  { 
    Console.Write ("{0}{1}", prefix,msg); 
  } 
} 


public class case_portfolio_1
{

    public static void Main (String[] args)
    {

        const int n = 3;

        // Since the value infinity is never used, we define
        // 'infinity' symbolic purposes only
        double infinity = 0.0;
        double gamma = 0.05;
        double[]   mu = {0.1073,  0.0737,  0.0627};
        double[,] GT={
            {0.1667,  0.0232,  0.0013},
            {0.0000,  0.1033, -0.0022},
            {0.0000,  0.0000,  0.0338}
        };
        double[] x0 = {0.0, 0.0, 0.0};
        double   w = 1.0;

        int numvar = 2*n +1;
        int numcon = n+1;


        //Offset of variables into the API variable.  
        int offsetx = 0;
        int offsets = n;
        int offsett = n+1;

        // Make mosek environment. 
        using (mosek.Env env = new mosek.Env()) 
        {   
            // Create a task object. 
            using (mosek.Task task = new mosek.Task(env,0,0)) 
            { 
                // Directs the log task stream to the user specified 
                // method msgclass.streamCB 
                task.set_Stream (mosek.streamtype.log, new msgclass ("")); 

                
                //Constraints.
                task.appendcons(numcon);
                for( int i=1; i<=n; ++i)
                    {
                        w+=x0[i-1];
                        task.putconbound(i, mosek.boundkey.fx, infinity,infinity);
                        task.putconname(i,"GT["+i+"]");
                    }
                task.putconbound(0, mosek.boundkey.fx,w,w);
                task.putconname(0,"budget");

                //Variables.
                task.appendvars(numvar);

                int[] xindx={offsetx+0,offsetx+1,offsetx+2};
                task.putclist(xindx,mu);

                for( int i=0; i<n; ++i)
                    {
                        for( int j=i; j<n;++j)
                            task.putaij(i+1,offsetx+j, GT[i,j]); 

                        task.putaij(i+1, offsett+i,-1.0);

                        task.putvarbound(offsetx+i, mosek.boundkey.lo,infinity,infinity);

                        task.putvarname(offsetx+i,"x["+(i+1)+"]");
                        task.putvarname(offsett+i,"t["+(i+1)+"]");
                        task.putvarbound(offsett+i, mosek.boundkey.fr,infinity,infinity);
                    }
                task.putvarbound(offsets, mosek.boundkey.fx,gamma,gamma);
                task.putvarname(offsets,"s");

                double[] e={1.0,1.0,1.0};
                task.putarow(0,xindx,e);

                //Cones.
                int[] csub = {offsets,offsett+0,offsett+1,offsett+2};
                task.appendcone( mosek.conetype.quad, 
                                 0.0, /* For future use only, can be set to 0.0 */
                                 csub);
                task.putconename(0,"stddev");

                /* A maximization problem */ 
                task.putobjsense(mosek.objsense.maximize);

                task.solutionsummary(mosek.streamtype.log);

                task.writedata("dump.opf");

                task.optimize();


                double expret=0.0,stddev=0.0;
                double[] xx = new double[numvar];

                task.getxx(mosek.soltype.itr,xx);
                for(int j=0; j<n; ++j)  
                    expret += mu[j]*xx[j+offsetx];

                Console.WriteLine("\neglected return {0:E} for gamma {1:E}\n",expret, xx[offsets]); 
            }
        }  
    }
}       
